### PR TITLE
Enhence generate proposal labels

### PR DESCRIPTION
**Pull request by xingyuanbu**

### CODE REVIEW

File: paddle/fluid/operators/detection/generate_proposal_labels_op.cc [hunk 1]
```diff
@@ -64,22 +65,25 @@ class GenerateProposalLabelsOp : public framework::OperatorWithKernel {
...
+    auto im_info_dims = ctx->GetInputDim("ImInfo");
 
     PADDLE_ENFORCE_EQ(rpn_rois_dims.size(), 2,
                       "The rank of Input(RpnRois) must be 2.");
-    PADDLE_ENFORCE_EQ(gt_classes_dims.size(), 1,
-                      "The rank of Input(GtClasses) must be 1.");
+    // PADDLE_ENFORCE_EQ(gt_classes_dims.size(), 2,
+    //                   "The rank of Input(GtClasses) must be 2.");
+    // PADDLE_ENFORCE_EQ(is_crowd_dims.size(), 2,
+    //                   "The rank of Input(IsCrowd) must be 2.");
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/generate_proposal_labels_op.cc]: Please remove the commented lines.

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 2]
```diff
@@ -43,10 +49,22 @@ class RpnTargetAssignOp : public framework::OperatorWithKernel {
...
+    auto anchor_dims = ctx->GetInputDim("Anchor");
+    auto gt_boxes_dims = ctx->GetInputDim("GtBoxes");
+    auto is_crowd_dims = ctx->GetInputDim("IsCrowd");
+    auto im_info_dims = ctx->GetInputDim("ImInfo");
+    PADDLE_ENFORCE_EQ(anchor_dims.size(), 2,
+                      "The rank of Input(Anchor) must be 2.");
+    PADDLE_ENFORCE_EQ(gt_boxes_dims.size(), 2,
+                      "The rank of Input(GtBoxes) must be 2.");
+    // PADDLE_ENFORCE_EQ(is_crowd_dims.size(), 2,
+    //                   "The rank of Input(IsCrowd) must be 2.");
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: remove this line.

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 3]
```diff
@@ -59,198 +77,385 @@ class RpnTargetAssignOp : public framework::OperatorWithKernel {
...
+      T diff = std::abs(value - gt_to_anchor_max_data[j]);
+      if (diff < epsilon) {
+        is_anchors_with_max_overlap = true;
+        break;
+      }
+    }
+    bool is_anchor_great_than_thresh =
+        (anchor_to_gt_max_data[i] >= rpn_positive_overlap);
+    if (is_anchors_with_max_overlap || is_anchor_great_than_thresh) {
+      // target_label[i] = 1;
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: remove this line.

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc [hunk 4]
```diff
@@ -59,198 +77,385 @@ class RpnTargetAssignOp : public framework::OperatorWithKernel {
...
+  }
 
-    auto lod = dist_t->lod().back();
-    int64_t batch_num = static_cast<int64_t>(lod.size() - 1);
-    int64_t anchor_num = dist_t->dims()[1];
-    PADDLE_ENFORCE_EQ(anchor_num, anchor_t->dims()[0]);
+  int bg_num = rpn_batch_size_per_im - fg_num;
+  for (int64_t i = 0; i < anchor_num; ++i) {
+    if (anchor_to_gt_max_data[i] < rpn_negative_overlap) {
+      // target_label[i] = 0;
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/rpn_target_assign_op.cc]: remove this line.

File: paddle/fluid/operators/detection/generate_proposal_labels_op.cc (hunk 1 above)
Comment (xingyuanbu) [reply paddle/fluid/operators/detection/generate_proposal_labels_op.cc]: revised

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc (hunk 3 above)
Comment (xingyuanbu) [reply paddle/fluid/operators/detection/rpn_target_assign_op.cc]: revised

File: paddle/fluid/operators/detection/rpn_target_assign_op.cc (hunk 4 above)
Comment (xingyuanbu) [reply paddle/fluid/operators/detection/rpn_target_assign_op.cc]: revised

**Review by qingqing01 — APPROVED** [review]

**Comment by qingqing01** [comment]
@panyx0718 @typhoonzero There are API changes, please help to review. Thanks!

File: paddle/fluid/API.spec [hunk 5]
```diff
@@ -305,9 +305,9 @@ paddle.fluid.layers.target_assign ArgSpec(args=['input', 'matched_indices', 'neg
...
 paddle.fluid.layers.anchor_generator ArgSpec(args=['input', 'anchor_sizes', 'aspect_ratios', 'variance', 'stride', 'offset', 'name'], varargs=None, keywords=None, defaults=(None, None, [0.1, 0.1, 0.2, 0.2], None, 0.5, None))
-paddle.fluid.layers.generate_proposal_labels ArgSpec(args=['rpn_rois', 'gt_classes', 'gt_boxes', 'im_scales', 'batch_size_per_im', 'fg_fraction', 'fg_thresh', 'bg_thresh_hi', 'bg_thresh_lo', 'bbox_reg_weights', 'class_nums'], varargs=None, keywords=None, defaults=(256, 0.25, 0.25, 0.5, 0.0, [0.1, 0.1, 0.2, 0.2], None))
+paddle.fluid.layers.generate_proposal_labels ArgSpec(args=['rpn_rois', 'gt_classes', 'is_crowd', 'gt_boxes', 'im_info', 'batch_size_per_im', 'fg_fraction', 'fg_thresh', 'bg_thresh_hi', 'bg_thresh_lo', 'bbox_reg_weights', 'class_nums', 'use_random'], varargs=None, keywords=None, defaults=(256, 0.25, 0.25, 0.5, 0.0, [0.1, 0.1, 0.2, 0.2], None, True))
```
Comment (typhoonzero) [inline paddle/fluid/API.spec:310]: should be aware that adding placement argument `is_crowd` will break API compatibility. same as above APIs.

Comment (qingqing01) [reply paddle/fluid/API.spec:310]: @typhoonzero Thanks!  This API is never used by others yet, it is used in faster-rcnn, which is not released yet.

**Review by typhoonzero — APPROVED** [review]
Approve for API change!

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**